File: leetcode/2024_8_august/c++/13_combination_sum_ii.cpp
// TIME COMPEXITY
// O(2^n)

// SPACE COMPEXITY
// O(n)

// CODE

class Solution {
public:
    void solve(int ind, vector<int>& arr, int target, vector<vector<int> > &ans, vector<int> &temp){
        if(target == 0){
            ans.push_back(temp);
            return ;
        }

        for(int i = ind; i < arr.size(); i++){
            if(i > ind && arr[i] == arr[i-1]) continue;
            if(arr[i] > target) break;
            temp.push_back(arr[i]);
            solve(i+1,arr,target-arr[i],ans,temp);
            temp.pop_back();
        }
    }
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<vector<int> > ans;
        vector<int> temp;
        solve(0,candidates,target,ans,temp);
        return ans;

    }
};

File: leetcode/2024_8_august/python/13_combination_sum_ii.py
# CODE 

from typing import List

class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        candidates.sort()
        
        def solve(idx, path, sum, res, candidates):
            if sum <= 0:
                if sum == 0:
                    res.append(path)
                return 
            for i in range(idx, len(candidates)):
                if i == idx or candidates[i] != candidates[i - 1]:
                    solve(i + 1, path + [candidates[i]], sum - candidates[i], res, candidates)

        solve(0, [], target, res, candidates)
        return res         